macrosynergy/pnl/contract_signals.py

```diff
@@ -0,0 +1,621 @@
+    update_df,
+    reduce_df,
+    estimate_release_frequency,
+)
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def _check_scaling_args(
+    ctypes: List[str],
+    cscales: Optional[List[Union[Number, str]]] = None,
+    csigns: Optional[List[int]] = None,
+    hbasket: Optional[List[str]] = None,
+    hscales: Optional[List[Union[Number, str]]] = None,
+    hratios: Optional[str] = None,
+) -> Tuple[Any, Any, Any, Any, Any]:
+
+    ## Check cscales and csigns
+    if cscales is not None:
+        # check that the number of scales is the same as the number of ctypes
+        if len(cscales) != len(ctypes):
+            raise ValueError("`cscales` must be of the same length as `ctypes`")
+        if not all([isinstance(x, (str, Number)) for x in cscales]):
+            raise TypeError("`cscales` must be a List of strings or numerical values")
+    else:
+        cscales: List[Number] = [1.0] * len(ctypes)
+
+    if csigns is not None:
+        # check that the number of signs is the same as the number of ctypes
+        if len(csigns) != len(ctypes):
+            raise ValueError("`csigns` must be of the same length as `ctypes`")
+        if not all([isinstance(x, int) for x in csigns]):
+            raise TypeError("`csigns` must be a List of integers")
+        if not all([x in [-1, 1] for x in csigns]):
+            warnings.warn(
+                "`csigns` must be a List of -1 or 1, coercing to -1 or 1",
+                UserWarning,
+            )
+            csigns: List[int] = [x / abs(x) for x in csigns]
+    else:
+        csigns: List[int] = [1] * len(ctypes)
+
+    ## Check hbasket and hscales
+    if hbasket is not None:
+        if not (bool(hratios) and bool(hscales)):
+            raise ValueError(
+                "`hratios` and `hscales` must be provided if `hbasket` is provided"
+            )
+        else:
+            # check that the number of scales is the same as the number of hbasket
+            if len(hscales) != len(hbasket):
+                raise ValueError("`hscales` must be of the same length as `hbasket`")
+            if not all([isinstance(x, (str, Number)) for x in hscales]):
+                raise TypeError(
+                    "`hscales` must be a List of strings or numerical values"
+                )
+
+    return cscales, csigns, hbasket, hscales, hratios
+
+
+def _check_estimation_frequency(df_wide: pd.DataFrame, rebal_freq: str) -> pd.DataFrame:
+    """
+    Check the timeseries to see if the estimated frequency matches the actual frequency.
+
+    Parameters
+    ----------
+    df_wide : pd.DataFrame
+        dataframe in wide format with the contract signals.
+    est_freq : str
+        the estimated frequency of the contract signals.
+
+    Raises
+    ------
+    ValueError
+        if the estimated frequency does not match the actual frequency.
+
+    Returns
+    -------
+    pd.DataFrame
+        dataframe with the estimated frequency.
+    """
+
+    return
```

Comment:
Why do we have an empty return here?

Reply:
The function is currently ununsed and the backend functionality was not working as expected at the time of implementation.
This logic was fixed a few weeks ago ([PR 2133](https://github.com/macrosynergy/macrosynergy/pull/2133)), so the function should now run as the docstring says